src/pseudo/PspUpf.jl

```diff
@@ -0,0 +1,220 @@
+                                     " features/quantities: $unsupported")
+
+    Zion        = Int(pseudo["header"]["z_valence"])    # Pseudo-ion charge
+    rgrid       = pseudo["radial_grid"]                 # Radial grid
+    drgrid      = pseudo["radial_grid_derivative"]      # Integration coeff.s
+    lmax        = pseudo["header"]["l_max"]             # Maximum angular momentum channel
+    nproj       = pseudo["header"]["number_of_proj"]    # Number of BKB projectors
+    nwfc        = pseudo["header"]["number_of_wfc"]     # Number of pseudo-atomic ψ
+    vloc        = pseudo["local_potential"] ./ 2        # Local potential (Ry -> Ha)
+    vloc_interp = linear_interpolation((rgrid,), vloc)  # Interp.or for the local potential
+    IT = typeof(vloc_interp)
+
+    # Kleinman-Bylander β projectors projs[l+1][i][ir]
+    # NB: UPFs store rβ, not just β
+    projs = [Vector[] for _ = 0:lmax]   # Projectors on the grid
+    projs_interp = [IT[] for _ = 0:lmax]  # Interpolators for the non-local projectors
+    for i = 1:nproj
+        proj_data = pseudo["beta_projectors"][i]
+        l = proj_data["angular_momentum"]
+        proj = proj_data["radial_function"] ./ 2  # Ry -> Ha
+        ir_start = iszero(rgrid[1]) ? 2 : 1  # Some grids start at 0., so (rβ)[1]/r[1] is undefined
+        ir_cut = length(proj)  # Projectors are cut off by the UPF parser @ ir_cut given by the UPF
+        proj_interp = linear_interpolation((rgrid[ir_start:ir_cut],), proj[ir_start:ir_cut])
+        push!(projs[l+1], proj)
+        push!(projs_interp[l+1], proj_interp)
+    end
+
+    # β-projector coupling coefficients h[l+1][i,j] (also called e_{KB}, D_{ij})
+    h = Matrix[]
+    n = 1
+    for l = 0:lmax
+        nproj_l = length(projs[l+1])
+        Dij_l = pseudo["D_ion"][n:n+nproj_l-1, n:n+nproj_l-1] .* 2  # 1/Ry -> 1/Ha
+        push!(h, Dij_l)
+        n += nproj_l
+    end
+
+    # Pseudo-atomic wavefunctions
+    # Currently not used; can be used for initializing the starting wavefunction and as
+    # projectors for projected densities of states, projected wavefunctions, and DFT+U(+V)
+    pswfcs = Vector{Vector}(undef, nwfc)
+    pswfc_ang_moms = Vector(undef, nwfc)
+    pswfc_occs = Vector(undef, nwfc)
+    for i = 1:nwfc
+        pswfc_data = pseudo["atomic_wave_functions"][i]
+        pswfcs[i] = pswfc_data["radial_function"]
+        pswfc_ang_moms[i] = pswfc_data["angular_momentum"]
+        pswfc_occs[i] = pswfc_data["occupation"]
+    end
+
+    # Currenctly used; can be used for initializing the starting guess density.
+    rhoatom = pseudo["total_charge_density"]
+
+    # Useful precomputed quantities (q-independent parts of Fourier transform integrands)
+    rvlocpzdr = (rgrid .* vloc .+ Zion) .* drgrid
+    r2projsdr = [Vector[] for _ = 0:lmax]
+    for l = 0:lmax
+        for proj in projs[l+1]
+            push!(r2projsdr[l+1],
+                  proj .* view(rgrid, 1:length(proj)) .* view(drgrid, 1:length(proj)))
+        end
+    end
```

Comment:
Instead of `view` use `@views` in front of the loops. Again this calls for list comprehensions or maps :smile:.